fix: file keyring fails to add/import/export keys when input is not stdin (fix #9566)

daeMOn63 opened the pull request:
## Description

Add a test case to reproduce the issue described in #9566. The test currently fails, and I've pointed some possible solutions over https://github.com/cosmos/cosmos-sdk/issues/9566#issuecomment-889281861. But I feel this needs more works in order to provide a more robust solution. I'll keep poking at better options, but taking any pointers if anyone has ideas.

amaurym commented on client/keys/add_test.go:
```diff
@@ -114,3 +116,47 @@ func Test_runAddCmdBasic(t *testing.T) {
 	require.Error(t, cmd.ExecuteContext(ctx))
 }
+
+func TestAddRecoverFileBackend(t *testing.T) {
```
afaict, this bug is also present on master. Maybe let's rebase this PR to master, and create backports to 0.42.x?

daeMOn63 replied:
indeed. Thanks, I'll attempt a fix and update PR target at this time!

amaurym reviewed (commented):
Wow, thanks for this debugging @daeMOn63! Really appreciate it.

I read your analysis in the issue. IMO your solution is okay, we can use it for now as a patch as it solves the immediate bug, and maybe put a TODO to improve it in the future.

daeMOn63 commented:
@amaurym I've updated the PR with an actual fix, and rebased & retargeted on master. Also added a couple of tests to check for this issue on other impacted commands I could spot. 

I may have found a way to not require applications to be updated by creating the needed `*bufio.Reader` directly from inside the `context.WithInput` function. 

Not sure about the backport process - is there anything I need to do ?

alessio reviewed (approved):
Good to go

amaurym commented on client/context.go:
```diff
@@ -68,7 +69,10 @@ func (ctx Context) WithKeyringOptions(opts ...keyring.Option) Context {
-	ctx.Input = r
+	// convert to a bufio.Reader to have a shared buffer between the keyring and the
+	// the Commands, ensuring a read from one advance the read pointer for the other.
+	// see #9566.
```
```suggestion
	// see https://github.com/cosmos/cosmos-sdk/issues/9566.
```

amaurym reviewed (approved):
tACK only on macos though, there's no regression on macos.

amaurym commented:
@mergifyio backport release/v0.42.x

amaurym commented:
@mergifyio backport release/v0.43.x
